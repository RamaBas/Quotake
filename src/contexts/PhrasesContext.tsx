import React, { createContext, useState, useCallback, useContext } from 'react';
import type { Phrase, PhrasesContextType } from '../types/phrases';

const PhrasesContext = createContext<PhrasesContextType | undefined>(undefined);

export const PhrasesProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {

  // Falta desarrollar addPhrase para dejr de mockear esto
  const [phrases, setPhrases] = useState<Phrase[]>([
    { id: '1', text: 'Aquel que a buen árbol se arrima, buena sombra recibe' },
    { id: '2', text: 'La vida es lo que pasa mientras estás ocupado haciendo Challanges.' },
    { id: '3', text: 'La simplicidad es la máxima sofisticación.' },
    { id: '4', text: 'El único modo de hacer un gran trabajo es amar lo que haces.' },
  ]);

  const addPhrase = useCallback((text: string) => {
    setPhrases(prev => [
      ...prev,
      {
        id: Date.now().toString(),
        text,
      }
    ]);
  }, []);

  const deletePhrase = useCallback((id: string) => {
    setPhrases(prev => prev.filter(phrase => phrase.id !== id));
  }, []);

  const searchPhrases = useCallback((text: string) => {
    return text 
      ? phrases.filter(phrase => 
          phrase.text.toLowerCase().includes(text.toLowerCase())
        )
      : phrases;
  }, [phrases]);

  return (
    <PhrasesContext.Provider 
      value={{ phrases, addPhrase, deletePhrase, searchPhrases }}
    >
      {children}
    </PhrasesContext.Provider>
  );
};

export const usePhrases = (): PhrasesContextType => {
  const context = useContext(PhrasesContext);
  if (!context) {
    throw new Error('error en usePhrases');
  }
  return context;
};