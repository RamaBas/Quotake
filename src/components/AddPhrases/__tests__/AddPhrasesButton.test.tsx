import { render, screen, fireEvent } from '@testing-library/react';
import { usePhrases } from '../../../contexts/PhrasesContext';
import AddPhraseButton from '../AddPhrasesButton';

const mockAddPhrase = jest.fn();

beforeEach(() => {
  (usePhrases as jest.Mock).mockReturnValue({
    addPhrase: mockAddPhrase
  });
});

afterEach(() => {
  jest.clearAllMocks();
});

describe('AddPhraseButton', () => {
  it('renders the component', () => {
    const { asFragment } = render(<AddPhraseButton />);
    expect(screen.getByLabelText('add')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('opens modal when button is clicked', () => {
    render(<AddPhraseButton />);
    fireEvent.click(screen.getByLabelText('add'));
    expect(screen.getByText('addPhrase')).toBeInTheDocument();
  });

  it('shows error when submitting empty phrase', async () => {
    render(<AddPhraseButton />);
    fireEvent.click(screen.getByLabelText('add'));
    fireEvent.click(screen.getByText('saveButton'));
    
    expect(screen.getByText('emptyPhraseError')).toBeInTheDocument();
    expect(mockAddPhrase).not.toHaveBeenCalled();
  });
});