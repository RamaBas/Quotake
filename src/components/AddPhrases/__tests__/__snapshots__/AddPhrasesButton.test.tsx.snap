// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPhraseButton renders the component 1`] = `
<DocumentFragment>
  <button
    aria-label="add"
    class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary css-1onxd2y-MuiButtonBase-root-MuiFab-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
      data-testid="AddIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
      />
    </svg>
  </button>
</DocumentFragment>
`;
